### PR TITLE
Fixed constraint lookup for multi-schema Postgres databases.

### PR DESCRIPTION
**Pull request by chrisspen**
When migrating indexes or constraints in PostgreSQL databases where the db user has access to multiple schemas, the query in DatabaseIntrospection.get_constraints will find constraints across every schemas, and not just the schema the current Django settings apply to.

This will result in a migration exception like:

    ValueError: Found wrong number (7) of constraints for myapp_mytable(some_col, some_other_col)

because it's expecting to find a single constraint (the old one) and it's seeing constraints from every schema the current user has access to.

The fix is fairly simple, which just involves looking up the current schema, and then filtering the constraint lookup by that schema.

### CODE REVIEW

File: django/db/backends/postgresql/introspection.py [hunk 1]
```diff
@@ -207,10 +209,11 @@ def get_constraints(self, cursor, table_name):
                 LEFT JOIN pg_class c2 ON idx.indexrelid = c2.oid
                 LEFT JOIN pg_am am ON c2.relam = am.oid
                 LEFT JOIN pg_attribute attr ON attr.attrelid = c.oid AND attr.attnum = idx.key
+                JOIN pg_namespace nsp ON nsp.oid = c.relnamespace AND nsp.nspname = %s
```
Comment (pope1ni) [inline django/db/backends/postgresql/introspection.py:212]: You could just do this:

```suggestion
                JOIN pg_namespace nsp ON nsp.oid = c.relnamespace AND nsp.nspname = current_schema()
```

File: django/db/backends/postgresql/introspection.py [hunk 2]
```diff
@@ -145,6 +145,8 @@ def get_constraints(self, cursor, table_name):
         one or more columns. Also retrieve the definition of expression-based
         indexes.
         """
+        cursor.execute("SELECT current_schema();")
+        current_schema = list(cursor.fetchall())[0][0]
```
Comment (pope1ni) [inline django/db/backends/postgresql/introspection.py:149]: Which avoids an extra query:
```suggestion
```

File: django/db/backends/postgresql/introspection.py [hunk 3]
```diff
@@ -207,10 +209,11 @@ def get_constraints(self, cursor, table_name):
                 LEFT JOIN pg_class c2 ON idx.indexrelid = c2.oid
                 LEFT JOIN pg_am am ON c2.relam = am.oid
                 LEFT JOIN pg_attribute attr ON attr.attrelid = c.oid AND attr.attnum = idx.key
+                JOIN pg_namespace nsp ON nsp.oid = c.relnamespace AND nsp.nspname = %s
                 WHERE c.relname = %s AND pg_catalog.pg_table_is_visible(c.oid)
             ) s2
             GROUP BY indexname, indisunique, indisprimary, amname, exprdef, attoptions;
-        """, [self.index_default_access_method, table_name])
+        """, [self.index_default_access_method, current_schema, table_name])
```
Comment (pope1ni) [inline django/db/backends/postgresql/introspection.py:216]:
```suggestion
        """, [self.index_default_access_method, table_name])
```

**Review by pope1ni — COMMENTED** [review]
All bug fixes require an accepted ticket. Please see the [contributing guidelines](https://docs.djangoproject.com/en/3.1/internals/contributing/).

We would also require a test to avoid any regressions in the future.

Moving on to the content of the proposed fix, I'm not sure that this is the right approach. [`current_schema()`](https://www.postgresql.org/docs/current/functions-info.html#FUNCTIONS-INFO-SESSION-TABLE) returns the first schema in the `search_path`, so relying on this is very likely to break in weird and wonderful ways.

When you create your ticket, please give as much detail as possible, via a simplified example, into the error you are seeing and what the structure of database looks like.

It sounds like you have constraints with identical names in different schemas. Can you just choose a different name for your constraint to ensure uniqueness?

**Comment by chrisspen** [comment]
Ok, thanks. I've created a bug report. https://code.djangoproject.com/ticket/32495

I suspected you might ask for a test, which is reasonable. However, because this addresses a multi-schema Postgres issue, how would I write a test for that? Is that case supported by the project's test framework?

What other approach would you recommend? Is see your point, but the only way this approach would break is if Django creates tables in different schemas. How likely is that to occur? My own case is rare enough, and I'm not even mixing schemas in my own logical sites. I'm only using schemas to isolate different databases.

**Comment by pope1ni** [comment]
> Is see your point, but the only way this approach would break is if Django creates tables in different schemas. How likely is that to occur?

Not very, but it also sets a bad precedent should we eventually support schemas properly.

> I'm only using schemas to isolate different databases.

I have to ask: Why are you not using different databases to isolate different databases? Using schemas for multi-tenanting your application is dangerous. Mistakes in setting `search_path` could lead to exposing data to the wrong users, or there is a risk of corrupting your data. Even the PostgreSQL [documentation](https://www.postgresql.org/docs/current/ddl-schemas.html#DDL-SCHEMAS-PATH) has warnings:

> The ability to create like-named objects in different schemas complicates writing a query that references precisely the same objects every time. It also opens up the potential for users to change the behavior of other users' queries, maliciously or accidentally. Due to the prevalence of unqualified names in queries and their use in PostgreSQL internals, adding a schema to search_path effectively trusts all users having CREATE privilege on that schema. When you run an ordinary query, a malicious user able to create objects in a schema of your search path can take control and execute arbitrary SQL functions as though you executed them.

Closing as per Mariusz' [comment](https://code.djangoproject.com/ticket/32495#comment:1).

**Comment by chrisspen** [comment]
>Why are you not using different databases to isolate different databases? Using schemas for multi-tenanting your application is dangerous. Mistakes in setting search_path could lead to exposing data to the wrong users, or there is a risk of corrupting your data. 

Yeah, I had similar reservations when I first considered this solution. In reality, it's not difficult to reliably set the search_path. It's specified via the database connection setting through the `OPTION` field like:

    'OPTIONS': {
        'options': '-c search_path=<schema name>'
    }

So to expose data by getting the search_path wrong would be to expose data by getting the username or database name values wrong, or using one tenant's connection parameters for a different tenant. Could I accidentally do that? In theory. But in theory, anything can go wrong. In practice, it's not an issue.

And the upside is that schemas are much cheaper and more lightweight that a full database, especially in a multi-tenant architecture. I host mostly in cloud environments, and the cost difference between creating a separate PostgreSQL database vs creating a separate PostgreSQL schema is significant.

That's the purpose behind [django-tenant-schemas](https://django-tenant-schemas.readthedocs.io/en/latest/), although I'm rolling my own implementation.

**Pull request closed without merging**